File: agent-framework-deepagents/server/services/skill_manager.py
import importlib.util
from pathlib import Path
from typing import Any, Optional

from server.models.oaf_types import SkillConfig


class SkillManager:
    """Skill 管理器

    动态加载 skills/ 下的 Python 脚本
    """

    def __init__(self, skills_dir: Path | str):
        self.skills_dir = Path(skills_dir)

    def load_all(self, skill_configs: list[SkillConfig]) -> list[dict]:
        loaded = []
        if not self.skills_dir.exists():
            return loaded

        for sc in skill_configs:
            if sc.source != "local":
                skill = self._load_remote(sc)
            else:
                skill = self._load_local(sc)
            if skill:
                loaded.append(skill)
        return loaded

    def _load_local(self, config: SkillConfig) -> Optional[dict]:
        skill_dir = self.skills_dir / config.name
        if not skill_dir.exists() or not skill_dir.is_dir():
            return None

        module = self._load_module(skill_dir)
        metadata = self._load_metadata(skill_dir, config)

        return {
            "name": config.name,
            "path": str(skill_dir),
            "metadata": metadata,
            "module": module,
            "config": config,
        }

    def _load_remote(self, config: SkillConfig) -> Optional[dict]:
        return {
            "name": config.name,
            "path": config.source,
            "metadata": {
                "name": config.name,
                "description": config.description,
                "version": config.version,
                "source": config.source,
            },
            "module": None,
            "config": config,
        }

    def _load_module(self, skill_dir: Path) -> Optional[Any]:
        scripts_dir = skill_dir / "scripts"
        if not scripts_dir.exists():
            return None

        for py_file in scripts_dir.glob("*.py"):
            if py_file.name.startswith("_"):
                continue
            try:
                module_name = f"skill_{skill_dir.name}_{py_file.stem}"
                spec = importlib.util.spec_from_file_location(module_name, str(py_file))
                if spec and spec.loader:
                    module = importlib.util.module_from_spec(spec)
                    spec.loader.exec_module(module)
                    return module
            except Exception:
                pass
        return None

    def _load_metadata(self, skill_dir: Path, config: SkillConfig) -> dict:
        skill_md = skill_dir / "SKILL.md"
        meta = {
            "name": config.name,
            "description": config.description,
            "version": config.version,
        }
        if skill_md.exists():
            content = skill_md.read_text(encoding="utf-8")
            if content.startswith("---"):
                parts = content.split("---", 2)
                if len(parts) >= 3:
                    import yaml
                    try:
                        fm = yaml.safe_load(parts[1].strip()) or {}
                        meta.update({
                            "name": fm.get("name", meta["name"]),
                            "description": fm.get("description", meta["description"]),
                            "license": fm.get("license", "MIT"),
                            "allowed_tools": fm.get("allowed-tools", []),
                        })
                        if "metadata" in fm:
                            meta.update(fm["metadata"])
                    except Exception:
                        pass
        return meta

    def invoke_skill(self, skill_name: str, input_data: str, loaded_skills: list[dict]) -> str:
        for skill in loaded_skills:
            if skill["name"] == skill_name:
                module = skill.get("module")
                if module is None:
                    return f"[Skill:{skill_name}] Remote skill - no local module loaded"
                if hasattr(module, "main"):
                    try:
                        result = module.main(input_data)
                        return str(result)
                    except Exception as e:
                        return f"[Skill:{skill_name}] Error: {e}"
                return f"[Skill:{skill_name}] Loaded but no main() entry point"
        return f"[Skill:{skill_name}] Not found"

    def get_skill_summaries(self, loaded_skills: list[dict]) -> list[dict]:
        summaries = []
        for skill in loaded_skills:
            meta = skill.get("metadata", {})
            summaries.append({
                "name": skill["name"],
                "description": meta.get("description", ""),
                "version": meta.get("version", ""),
                "has_module": skill.get("module") is not None,
            })
        return summaries

File: agent-framework/src/main/java/io/agentmanager/framework/service/SkillManager.java
package io.agentmanager.framework.service;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import io.agentmanager.framework.model.OafConfig;
import io.agentmanager.framework.model.OafConfig.SkillConfig;

@Service
public class SkillManager {
    private static final Logger log = LoggerFactory.getLogger(SkillManager.class);

    private final Path skillsDir;

    public SkillManager(OafConfig oafConfig, java.nio.file.Path configDir) {
        this.skillsDir = configDir.resolve("skills");
    }

    public List<SkillInfo> loadAll(List<SkillConfig> skillConfigs) {
        var loaded = new ArrayList<SkillInfo>();
        for (var sc : skillConfigs) {
            var info = loadSkill(sc);
            if (info != null) {
                loaded.add(info);
            }
        }
        return loaded;
    }

    private SkillInfo loadSkill(SkillConfig config) {
        var skillDir = skillsDir.resolve(config.name());
        if (!skillDir.toFile().exists() || !skillDir.toFile().isDirectory()) {
            return null;
        }
        var metadata = loadMetadata(config, skillDir);
        return new SkillInfo(config.name(), skillDir.toString(), metadata, config);
    }

    private SkillMetadata loadMetadata(SkillConfig config, Path skillDir) {
        var skillMd = skillDir.resolve("SKILL.md");
        var meta = new SkillMetadata(config.name(), config.description(), config.version(), List.of());
        if (skillMd.toFile().exists()) {
            try {
                var content = java.nio.file.Files.readString(skillMd);
                if (content.startsWith("---")) {
                    var parts = content.split("---", 3);
                    if (parts.length >= 3) {
                        var yaml = new org.yaml.snakeyaml.Yaml();
                        @SuppressWarnings("unchecked")
                        var fm = (java.util.Map<String, Object>) yaml.load(parts[1].trim());
                        if (fm != null) {
                            @SuppressWarnings("unchecked")
                            var allowed = (java.util.List<String>) fm.getOrDefault("allowed-tools", List.of());
                            meta = new SkillMetadata(
                                (String) fm.getOrDefault("name", meta.name()),
                                (String) fm.getOrDefault("description", meta.description()),
                                (String) fm.getOrDefault("version", meta.version()),
                                allowed
                            );
                        }
                    }
                }
            } catch (Exception e) {
                log.warn("Failed to load SKILL.md for {}: {}", config.name(), e.getMessage());
            }
        }
        return meta;
    }

    public List<java.util.Map<String, Object>> getSkillSummaries(List<SkillInfo> loaded) {
        var summaries = new ArrayList<java.util.Map<String, Object>>();
        for (var skill : loaded) {
            summaries.add(java.util.Map.of(
                "name", skill.name(),
                "description", skill.metadata().description(),
                "version", skill.metadata().version()
            ));
        }
        return summaries;
    }

    public record SkillInfo(String name, String path, SkillMetadata metadata, SkillConfig config) {}
    public record SkillMetadata(String name, String description, String version, List<String> allowedTools) {}
}
